### PR TITLE
Add spec of thread-safety accessing zero-arg method

### PR DESCRIPTION
**Pull request by ms-ati**
See discussion in #224

### CODE REVIEW

**Comment by ms-ati** [comment]
cc @JacobEvelyn @jemmaissroff this is a pending spec for the thread-safety issue identified and fixed by @jeremyevans in #224 -- appreciate your review, so we can merge this pending spec before the fix?

File: spec/thread_safety_spec.rb
Comment (jemmaissroff) [inline spec/thread_safety_spec.rb:13]: I think the `map` is superfluous with `Array.new(2)` as well. `Array#new` can optionally take a block, so we could do:

```ruby
Array.new(2) do
   ...
end
```

File: spec/thread_safety_spec.rb
Comment (jemmaissroff) [inline spec/thread_safety_spec.rb:17]: Do we need to define `threads` at all, or could the above become:

```ruby
thread_return_values = Array.new(2) do
  Thread.new { instance.return_first_thread_id }
end.map(&:value)
```

File: spec/thread_safety_spec.rb
Comment (jemmaissroff) [inline spec/thread_safety_spec.rb:7]: I think this whole block could be simplified to

```ruby
let(:thread_return_values) do
  instance = class_with_memo.new

  Array.new(2) do
    Thread.new { instance.return_first_thread_id }
  end.map(&:value)
end
```

Rationale:
* Now that it's deterministic, we don't need the `1000.times` block
* `Array#new` takes an optional block, so we don't need the `#map` there too
* We don't need the break condition anymore
* No need to define `threads` or `thread_return_values`

File: spec/thread_safety_spec.rb
Comment (jemmaissroff) [inline spec/thread_safety_spec.rb:42]: I think this test will still fail even with @jeremyevans's fix. The rationale is that both threads will still hit this line:
```ruby
ret = @_memo_wise[#{index}] = #{original_memo_wised_name}
```
at roughly the same time, and still therefore get different values from their respective `return_first_thread_id` calls.

The case that Jeremy's PR has fixed is the nil case - neither will have a `nil` value. Do we also want to address this first case and ensure that both threads have the same memoized values? If so, might we need to use a Mutex or something like that (presumably at a performance cost in the non-memoized case) to lock the `@_memo_wise` cache?

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: spec/thread_safety_spec.rb
Comment (jeremyevans) [reply spec/thread_safety_spec.rb:42]: I think it's expected that any method that is cached should return the same value no matter which thread is calling it.  The test breaks that assumption.  Maybe the test should just check that the value returned is not nil, because in the failure case before my fix, `nil` should be returned if the race is lost.

File: spec/thread_safety_spec.rb
Comment (ms-ati) [reply spec/thread_safety_spec.rb:7]: @jemmaissroff while it is true that in the current code and test, we can indeed reliably reproduce the race on the first iteration, are you sure it's best to leave that assumption in place for an on-going regression test? In general, when leaving tests for previous race conditions in code, my norm has been to leave enough iterations up to 1 or 2 seconds of test time, so that if we every have a case that occurs probabilistically, which is the norm, we will eventually catch it in CI.

I'm open to thoughts either way! I just don't want my "success" here in triggering on the first iteration to have too much influence in how we think about designing race condition regression tests, if that makes sense.

File: spec/thread_safety_spec.rb
Comment (ms-ati) [reply spec/thread_safety_spec.rb:42]: 😝  I was hoping nobody would see that until we went to enable this test on Jeremy's PR

Comment (ms-ati) [reply spec/thread_safety_spec.rb:42]: (I was hoping to drive an explicit decision on this point within the context of his proposed fix)

What if I split the assertions into two, separate pending specs? And we can discuss whether Jeremy's PR should close one of them, and the other can be addressed in a separate PR, or converted to a documentary spec that shows that we expect possibly multiple values when the memoization is contended? (which to be clear is an acceptable design decision, that we simply need to document if so)

Comment (ms-ati) [reply spec/thread_safety_spec.rb:42]: @jeremyevans looks like we "typed past each other" here, as you can see in the timestamps of our comments above ;)

So to playback what we are all saying: this test has two distinct expectations.

Split-tracking:

- (a) Test of accidental `nil`
  - Proposal: Separate to its own test + remove `pending` as part of Jeremy's PR that fixes it?

- (b) Test of same return value when the memoization is contended
  - Option 1: Remove this test entirely
  - Option 2: Change test to be documentary - that this issue exists and should be expected
  - Option 3: Separate `pending` test - and plan to introduce a separate fix per Jeremy's description of the expectation above that we should not return different values when memoization is contended

Re-reading what everyone wrote above, what do we think about (a) separating, and (b-3) leaving a separate pending test in place to address the contended values issue in a different PR?

Comment (jeremyevans) [reply spec/thread_safety_spec.rb:42]: I personally see no value in testing behavior for a non-idempotent method.  Nobody should be using caching for non-idempotent methods.  So I would only check that the race condition is fixed (the test for non-nil results on a method that returns non-nil).

Comment (ms-ati) [reply spec/thread_safety_spec.rb:42]: I am totally fine removing this case entirely for now. 

I suspect that, per #184, we must eventually document what happens in contended cases, for example:

- Queries of external state
  - Should reading data in from a web request, database query, or file be memoized?
  - If so, what is the actual vs expected behavior, when contended queries generate different query results?

Unless I hear back differently from @jemmaissroff or @JacobEvelyn, I will plan to remove this test case for now, and we can discuss under #184.

Comment (JacobEvelyn) [reply spec/thread_safety_spec.rb:42]: Thanks for the thoughts, @ms-ati @jeremyevans @jemmaissroff ! I don't have strong feelings on what the behavior should be for nondeterministic methods, so I'm happy to defer to you three on either providing a stronger consistency guarantee or simply changing the test to confirm that the results are not `nil`. I agree with @ms-ati that we should document the behavior wherever we land.

File: spec/thread_safety_spec.rb
Comment (JacobEvelyn) [reply spec/thread_safety_spec.rb:7]: I don't think I have strong feelings here. I see the argument for letting this run for a while to catch nondeterministic failures, but I also think @jemmaissroff 's point about simplifying this if the failure _is_ deterministic makes sense. I'd be happy with either:

- simplifying the test as @jemmaissroff suggests
- keeping the code as-is and adding a comment explaining the rationale that @ms-ati outlined above

Comment (ms-ati) [reply spec/thread_safety_spec.rb:35]: Request for re-review @jemmaissroff -- I've extracted code and added comments to provide a rationale for time-limited probabilistic testing strategy for race conditions. What do you think?

File: spec/thread_safety_spec.rb
Comment (ms-ati) [reply spec/thread_safety_spec.rb]: Per discussion above, I've removed the test of same return value under contended memoization, we can discuss this separately.

File: spec/thread_safety_spec.rb
Comment (jemmaissroff) [reply spec/thread_safety_spec.rb:37]: Since this race condition reproduction is deterministic, I personally was confused by the repetition here.

Although I can admire the elegance of `check_repeatedly`, it seems to me to be over-engineering in this case and solving for a problem that we don't currently have. If we at some future point either have a race condition which we need to repeatedly run to reproduce, or this race condition for some reason requires repetition, this method looks perfect.

In the absence of this however, I think the repetition here overcomplicates and makes it harder to reason about a test which only requires one iteration to accomplish its goal.

Comment (ms-ati) [reply spec/thread_safety_spec.rb:37]: @jemmaissroff Ok, it seems like you would advocate for not having the practice for time-limited probabilistic testing in place until we have discovered a race condition which manifests probabilitically?

I thought having this in place in this new file `thread_safety.rb` would lower the work required for us to add tests for other race conditions, as we work to land thread-safety and documentation of thread contention behavior across the project.

So is there a natural tension between these interests?

1. Avoid engineering test code for probabilistic race conditions until we have one at hand
2. Make it easier to detect and add tests for probabilistic race conditions

Comment (jemmaissroff) [reply spec/thread_safety_spec.rb:37]: I might be missing context here - do we have plans to add any other race condition tests? Or to do more work on creating more race condition tests?

Comment (ms-ati) [reply spec/thread_safety_spec.rb:37]: _To be clear: I am happy removing this machinery for now!_

I am hoping to reach alignment on whether there are distinct interests at play in this decision? Are we choosing to serve 1 over 2, or are saying that 2 is not a valid interest ("over-engineering" and "overcomplicates")?

Comment (jemmaissroff) [reply spec/thread_safety_spec.rb:37]: Yes, I am aligned with removing it.

I'm trying to understand whether 2 is a valid interest. My assumption earlier in this thread was that we don't have plans to add other race condition tests, and so under that assumption I suggested that it is overcomplicating our codebase. When you outlined the interests, it made me realize I had made this assumption, and so I'm looking for clarity on whether we do have plans to add more race condition tests or not. 

If we don't have plans to do this, I would say yes, accommodating a case that doesn't exist is not a valid interest.

If we do have plans to do this, I'm interested in hearing more about this before I feel equipped to judge which interest we should prioritize.

Comment (ms-ati) [reply spec/thread_safety_spec.rb:37]: Ah ok, thanks for clarifying @jemmaissroff. I had been under the assumption that this was the beginning of a series of work to eventually fully answer #184 -- and I had been imagining adding test coverage for potential races across all of the memoization code paths.

But I find it very persuasive that such tooling can wait at least one more PR, since this is the very first, none of those other plans might happen, and it isn't needed for just this PR.

I'll remove it for now and we can revisit if we do end up moving forward with thread-safety tests across all of the memoization paths.

Comment (ms-ati) [reply spec/thread_safety_spec.rb:5]: @jemmaissroff Removed as discussed, please re-review?

Comment (ms-ati) [reply spec/thread_safety_spec.rb:5]: Well 😆 the joke is on us @jemmaissroff @JacobEvelyn -- in fact, it appears that the issue is not actually deterministic, just appears often. So in fact, trying it once is not enough, as you can see from these CI failures: https://github.com/panorama-ed/memo_wise/runs/4231683800?check_suite_focus=true

Q: Shall I return the machinery for time-limited probabilistic tests then?

File: spec/thread_safety_spec.rb
Comment (JacobEvelyn) [inline spec/thread_safety_spec.rb:19]: Nitpick: I wonder if a more "Rubyesque" name would just be `first_thread_id`.

**Review by JacobEvelyn — APPROVED** [review]
LGTM with one optional change. Would like @jemmaissroff 's review on this too before merging.

File: spec/thread_safety_spec.rb
Comment (ms-ati) [reply spec/thread_safety_spec.rb]: Done!

**Review by jemmaissroff — APPROVED** [review]

**Pull request merged**
